### PR TITLE
Fixes long wait filter keyspace

### PR DESCRIPTION
**Pull request by rafael**
# Description

* The following PR fixes an issue with KeyspacesToWatch functionality. When this flag is provided, some keyspaces are removed. However, vtgates still try to wait for available tablet types in them. This PR makes sure that filter is respected.
* I also added some tests around `WaitForTablets` for existent and new functionality.

### CODE REVIEW

**Comment by deepthi** [comment]
do you know whether this also needs to be fixed on the new gateway (`TabletGateway`)?

File: go/vt/discovery/fake_healthcheck.go [hunk 1]
```diff
@@ -74,7 +77,30 @@ func (fhc *FakeHealthCheck) RegisterStats() {
 
 // WaitForAllServingTablets is not implemented.
 func (fhc *FakeHealthCheck) WaitForAllServingTablets(ctx context.Context, targets []*querypb.Target) error {
-	return nil
+	if !fhc.ShouldWaitForTablets {
```
Comment (rafael) [inline go/vt/discovery/fake_healthcheck.go:80]: all the other tests rely on the fake implementation. We could update them in a different PR to simplify this. For now added this extra option.

File: go/vt/discovery/healthcheck.go [hunk 2]
```diff
@@ -615,8 +615,30 @@ func (hc *HealthCheckImpl) WaitForAllServingTablets(ctx context.Context, targets
...
+	// Let's remove from the target shards that are not in the keyspaceToWatch list.
+	for i, target := range targets {
+		keepTarget := false
+		for _, keyspaceToWatch := range KeyspacesToWatch {
+			if target.Keyspace == keyspaceToWatch {
+				keepTarget = true
+			}
+		}
+		if !keepTarget {
+			log.Infof("removing the following target as is not part of keyspaces_to_watch: %v", target)
```
Comment (deepthi) [inline go/vt/discovery/healthcheck.go:631]:
```suggestion
			log.Infof("removing target that is not part of keyspaces_to_watch: %v", target)
```

File: go/vt/discovery/healthcheck.go [hunk 3]
```diff
@@ -648,6 +670,11 @@ func (hc *HealthCheckImpl) waitForTablets(ctx context.Context, targets []*query.
 		select {
 		case <-ctx.Done():
 			timer.Stop()
+			for _, target := range targets {
+				if target != nil {
+					log.Infof("couldn't find tablets for the following target: %v", target)
```
Comment (deepthi) [inline go/vt/discovery/healthcheck.go:675]:
```suggestion
					log.Infof("couldn't find tablets for target: %v", target)
```

File: go/vt/discovery/legacy_tablet_stats_cache_wait.go [hunk 4]
```diff
@@ -84,6 +85,11 @@ func (tc *LegacyTabletStatsCache) waitForTablets(ctx context.Context, targets []
 		timer := time.NewTimer(waitAvailableTabletInterval)
 		select {
 		case <-ctx.Done():
+			for _, target := range targets {
+				if target != nil {
+					log.Infof("couldn't find tablets for the following target: %v", target)
```
Comment (deepthi) [inline go/vt/discovery/legacy_tablet_stats_cache_wait.go:90]:
```suggestion
					log.Infof("couldn't find tablets for target: %v", target)
```

File: go/vt/vtgate/discoverygateway.go [hunk 5]
```diff
@@ -207,6 +207,24 @@ func (dg *DiscoveryGateway) WaitForTablets(ctx context.Context, tabletTypesToWai
...
+	// Let's remove from the target shards that are not in the keyspaceToWatch list.
+	for i, target := range targets {
+		keepTarget := false
+		for _, keyspaceToWatch := range discovery.KeyspacesToWatch {
+			if target.Keyspace == keyspaceToWatch {
+				keepTarget = true
+			}
+		}
+		if !keepTarget {
+			log.Infof("removing the following target as is not part of keyspaces_to_watch: %v", target)
```
Comment (deepthi) [inline go/vt/vtgate/discoverygateway.go:223]:
```suggestion
			log.Infof("removing target that is not part of keyspaces_to_watch: %v", target)
```

**Review by deepthi — COMMENTED** [review]
The implementation feels a bit fragile. We are mutating the arguments passed to an exported function. Even though existing uses of that function are such that this won't cause problems, given that it _is_ an exported function, I feel that it is not a good idea. 
How about creating a new slice that consists of only the relevant keyspaces and returning that instead of operating directly on `targets`?

**Comment by rafael** [comment]
> The implementation feels a bit fragile. We are mutating the arguments passed to an exported function. Even though existing uses of that function are such that this won't cause problems, given that it _is_ an exported function, I feel that it is not a good idea.
> How about creating a new slice that consists of only the relevant keyspaces and returning that instead of operating directly on `targets`?

I think this is a good idea. I was on the fence because the existent code already mutates this same slice. I updated the code with your suggestion. We should probably do something similar [here](https://github.com/vitessio/vitess/blob/master/go/vt/discovery/healthcheck.go#L637)

**Review by deepthi — APPROVED** [review]
LGTM

**Pull request merged**